micropython/lora/sx1276.py

```diff
@@ -0,0 +1,620 @@
+        if isinstance(value, int):
+            value = bytearray([value])
+        try:
+            self._nss(0)
+            self._spi.read(1, reg | _WRITE_REG_BIT)
+            self._spi.write(value)
+        finally:
+            self._nss(1)
+        if __DEBUG__:
+            print("W {:#x} ==> {}".format(reg, binascii.hexlify(value)))
+            self._reg_read(reg)  # log the readback as well
+
+    def _reg_update(self, reg, update_mask, new_value):
+        # Note: this function has a race condition, if the register value changes between
+        # read and write then old_value will be stale
+        old_value = self._reg_read(reg)
+        if update_mask:
+            value = ((old_value & ~update_mask) & 0xFF) | (new_value & update_mask)
+            self._reg_write(reg, value)
+        return old_value
+
+    def _reg_read(self, reg):
+        buf = bytearray(1)
+        self._reg_readinto(reg, buf)
+        return buf[0]
+
+    def _reg_readinto(self, reg, buf):
+        try:
+            self._nss(0)
+            self._spi.read(1, reg)
+            self._spi.readinto(buf)
+            if __DEBUG__:
+                print("R {:#x} <== {}".format(reg, binascii.hexlify(buf)))
+        finally:
+            self._nss(1)
+
+    def _get_mode(self):
+        return self._reg_read(_REG_OPMODE) & _OPMODE_MODE_MASK
+
+    def _set_mode(self, mode, lora_en=True):
+        lora_val = _OPMODE_LONGRANGEMODE_LORA if lora_en else 0
+        old_value = self._reg_update(
+            _REG_OPMODE, _OPMODE_LONGRANGEMODE_LORA | _OPMODE_MODE_MASK, mode | lora_val
+        )
+        old_value &= _OPMODE_MODE_MASK
+        if old_value == _OPMODE_MODE_SLEEP:
+            time.sleep_ms(1)  # waking from sleep takes TS_OSC = 250us (typical)
+
+        if __DEBUG__:
+            print("Mode {} -> {} ({:#x})".format(old_value, mode, self._reg_read(_REG_OPMODE)))
+
+        return old_value
+
+    def _dio_isr(self, pin):
+        if pin not in self._irq_en:
+            # TODO: This check against _irq_en shouldn't really be necessary,
+            # but weirdly DIO1 (configured for RxTimeout) seems to go high
+            # during TX, for no apparent reason (it's also masked out in
+            # _IRQ_FLAGS_MASK). Pull-up/down on the pin doesn't change it. This
+            # is probably a driver bug, and once it's fixed then _irq_en can be
+            # fully removed as the modem IRQ mask should determine which IRQs
+            # are driven high
+            if __DEBUG__:
+                print("ISR ignored {}".format(pin))
+            return
+        if __DEBUG__:
+            print("ISR {}".format(pin))
+        if self._irq_callback:
+            self._irq_callback()
+
+    def receive(self, timeout_ms=None, rx_continuous=False):
+        # TODO: check for any pending TX
+        will_irq = self.start_receive(timeout_ms, rx_continuous)
+        while True:
+            if will_irq:
+                machine.idle()
```

Comment:
IMO this `idle()` should always be called, even if `will_irq` is False.  Because otherwise this is a very busy loop that consumes lots of power on the MCU side.

Reply:
@dpgeorge I guess this explains the high power consumption. I was worried on some ports this might get stuck in a WFI or similar if no interrupt was set, but I suppose there's always a timer tick or something to wake the CPU up again?

Reply:
Was reminded the original version of this code called `machine.lightsleep(10)` here. Maybe:

```python
if will_irq:
   machine.idle()
else:
   machine.lightsleep(1)
```

is a reasonable compromise? Max 1ms of lag on receive shouldn't be a problem even for a Class B LoRaWAN "Beacon" devices, I don't think.

Reply:
`machine.idle()` is perhaps a bit under-specified in the docs... we don't really have any tickless ports (*) so `idle()` should always wake after some short period.  Thus it would be safe to use it unconditionally here.

If you're concerned, then `time.sleep_ms(1)` would be an alternative.  Don't use `machine.lightsleep()`!  That will put it in a very low power consumption mode that usually disables USB.

(*) zephyr port is perhaps tickless, and `machine.idle()` there calls `k_yield()`.  I'm not sure exactly how that works if there are no interrupts to wake it up.